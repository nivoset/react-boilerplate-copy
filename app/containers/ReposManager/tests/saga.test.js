/**
 * Tests for ReposManager sagas
 */

import { put, takeLatest } from 'redux-saga/effects';

import { loadRepos, reposLoaded, repoLoadingError } from '../slice';

import githubData, { getRepos, addRepoOwnershipKey } from '../saga';

const username = 'username';

const repos = [
  {
    name: 'First repo',
    owner: {
      login: 'react-boilerplate',
    },
  },
  {
    name: 'Second repo',
    owner: {
      login: 'react-boilerplate',
    },
  },
];

const reposWithOwnershipKeys = addRepoOwnershipKey({ username, repos });

/* eslint-disable redux-saga/yield-effects */
describe('getRepos Saga', () => {
  let getReposGenerator;

  // We have to test twice, once for a successful load and once for an unsuccessful one
  // so we do all the stuff that happens beforehand automatically in the beforeEach
  beforeEach(() => {
    getReposGenerator = getRepos();

    const selectDescriptor = getReposGenerator.next().value;
    expect(selectDescriptor).toMatchSnapshot();

    const requestDescriptor = getReposGenerator.next(username).value;
    expect(requestDescriptor).toMatchSnapshot();

    const addRepoOwnershipKeyDescriptor = getReposGenerator.next(repos).value;
    expect(addRepoOwnershipKeyDescriptor).toMatchSnapshot();
  });

  it('should dispatch the reposLoaded action if it requests the data successfully', () => {
    const putDescriptor = getReposGenerator.next(reposWithOwnershipKeys).value;
    expect(putDescriptor).toEqual(
      put(reposLoaded({ repos: reposWithOwnershipKeys })),
    );
  });

  it('should call the repoLoadingError action if the response errors', () => {
    const putDescriptor = getReposGenerator.throw(new Error('Some error'))
      .value;
    expect(putDescriptor).toEqual(put(repoLoadingError()));
  });
});

describe('githubDataSaga Saga', () => {
  const githubDataSaga = githubData();

  it('should start task to watch for LOAD_REPOS action', () => {
    const takeLatestDescriptor = githubDataSaga.next().value;
    expect(takeLatestDescriptor).toEqual(takeLatest(loadRepos.type, getRepos));
  });
});
