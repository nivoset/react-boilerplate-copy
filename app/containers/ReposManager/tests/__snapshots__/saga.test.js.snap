// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getRepos Saga should call the repoLoadingError action if the response errors 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;

exports[`getRepos Saga should call the repoLoadingError action if the response errors 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "https://api.github.com/users/username/repos?type=all&sort=updated",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getRepos Saga should call the repoLoadingError action if the response errors 3`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      Object {
        "repos": Array [
          Object {
            "name": "First repo",
            "owner": Object {
              "login": "react-boilerplate",
            },
          },
          Object {
            "name": "Second repo",
            "owner": Object {
              "login": "react-boilerplate",
            },
          },
        ],
        "username": "username",
      },
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getRepos Saga should dispatch the reposLoaded action if it requests the data successfully 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;

exports[`getRepos Saga should dispatch the reposLoaded action if it requests the data successfully 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "https://api.github.com/users/username/repos?type=all&sort=updated",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getRepos Saga should dispatch the reposLoaded action if it requests the data successfully 3`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      Object {
        "repos": Array [
          Object {
            "name": "First repo",
            "owner": Object {
              "login": "react-boilerplate",
            },
          },
          Object {
            "name": "Second repo",
            "owner": Object {
              "login": "react-boilerplate",
            },
          },
        ],
        "username": "username",
      },
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;
